core/src/main/java/hudson/model/Descriptor.java

```diff
@@ -588,6 +602,78 @@ public T newInstance(StaplerRequest req, JSONObject formData) throws FormExcepti
     }
 
     /**
+     * Ensures that calls to {@link StaplerRequest#bindJSON(Class, JSONObject)} from {@link #newInstance(StaplerRequest, JSONObject)} recurse properly.
+     * {@code doConfigSubmit}-like methods will wind up calling {@code newInstance} directly
+     * or via {@link #newInstancesFromHeteroList(StaplerRequest, Object, Collection)},
+     * which consult any custom {@code newInstance} overrides for top-level {@link Describable} objects.
+     * But for nested describable objects Stapler would know nothing about {@code newInstance} without this trick.
+     */
+    @SuppressWarnings({"unchecked", "rawtypes"})
+    private static class NewInstanceBindInterceptor extends BindInterceptor {
+
+        private final BindInterceptor oldInterceptor;
+        private final Map<JSONObject,Void> processed = new IdentityHashMap<>();
```

Comment:
> Not clear what this does

Prevents a `StackOverflowError`. Noted in PR description.



> why it's reset before the `verifyNewInstance` call

It is not reset, what do you mean? The same `NewInstanceBindInterceptor` should be used for the duration of the `StaplerRequest` (except when `Node.reconfigure` is called, which causes another one to be used, but that should be harmless).